File: GenericTesting/GenericVBTesting/Models/DemandLocation.vb
Imports System.ComponentModel

Public NotInheritable Class DemandLocation
  Implements INotifyPropertyChanged

  'CONSTRUCTOR
  Public Sub New()
  End Sub
  Public Sub New(locationID As Integer, companyNbr As String, divisionNbr As String, branchNbr As String, branchName As String)
    Me.LocationID = locationID
    Me.CompanyNbr = companyNbr
    Me.DivisionNbr = divisionNbr
    Me.BranchNbr = branchNbr
    Me.BranchName = branchName
  End Sub


  'PROPERTIES
  Public Property LocationID As Integer

  Public Property CompanyNbr As String = String.Empty

  Public Property DivisionNbr As String = String.Empty

  Public Property BranchNbr As String = String.Empty

  Public Property BranchName As String = String.Empty

  Private _isUsed As Boolean

  Public Property IsUsed As Boolean
    Get
      Return _isUsed
    End Get
    Set(ByVal value As Boolean)
      _isUsed = value
      OnPropertyChanged(NameOf(IsUsed))
    End Set
  End Property

  'METHODS
  Public Overrides Function ToString() As String
    Return $"{CompanyNbr.Trim}-{DivisionNbr.Trim}-{BranchNbr.Trim} {BranchName.Trim}"
  End Function

  Public Overrides Function Equals(obj As Object) As Boolean
    Select Case True
      Case TypeOf obj Is Integer
        Return Nullable.Equals(Me.LocationID, CInt(obj))
      Case Else
        Dim othr As DemandLocation = TryCast(obj, DemandLocation)
        Return If(IsNothing(othr), False, Nullable.Equals(othr.LocationID, Me.LocationID))
    End Select
  End Function

  Public Overrides Function GetHashCode() As Integer
    Return LocationID.GetHashCode
  End Function

  Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

  Public Sub OnPropertyChanged(ByVal info As String)
    RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(info))
  End Sub

End Class

File: GenericTesting/GenericTesting/Models/DemandLocation.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharpDataAccess.Enterprise.Models
{
  public sealed class DemandLocation : INotifyPropertyChanged
  {

    //CONSTRUCTOR
    public DemandLocation()
    {
    }
    public DemandLocation(int locationID, string companyNbr, string divisionNbr, string branchNbr, string branchName)
    {
      this.LocationID = locationID;
      this.CompanyNbr = companyNbr;
      this.DivisionNbr = divisionNbr;
      this.BranchNbr = branchNbr;
      this.BranchName = branchName;
    }


    //PROPERTIES
    public int LocationID { get; set; }

    public string CompanyNbr { get; set; }

    public string DivisionNbr { get; set; }

    public string BranchNbr { get; set; }

    public string BranchName { get; set; }


    private bool _isUsed;
    public bool IsUsed
    {
      get { return _isUsed; }
      set
      {
        _isUsed = value;
        OnPropertyChanged(nameof(IsUsed));
      }
    }

    //METHODS
    public override string ToString()
    {
      return $"{CompanyNbr.Trim()}-{DivisionNbr.Trim()}-{BranchNbr.Trim()} {BranchName.Trim()}";
    }

    public override bool Equals(object obj)
    {
     if (obj.GetType() == typeof(int))
      {
        return Nullable.Equals(this.LocationID, Convert.ToInt32(obj));
      }                                                               
     else
      {
        DemandLocation othr = obj as DemandLocation;
        return (othr == null) ? false : Nullable.Equals(othr.LocationID, this.LocationID);
      }

    }

    public override int GetHashCode()
    {
      return LocationID.GetHashCode();
    }

    public event PropertyChangedEventHandler PropertyChanged;

    public void OnPropertyChanged(string info)
    {
      PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(info));
    }

  }
}
